L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,938 @@
+  }
+
+  return lpgbt_ids;
+}
+
+GlobalPoint HGCalTriggerGeometryV9Imp3::getTriggerCellPosition(const unsigned trigger_cell_det_id) const {
+  unsigned det = DetId(trigger_cell_det_id).det();
+  // Position: barycenter of the trigger cell.
+  Basic3DVector<float> triggerCellVector(0., 0., 0.);
+  const auto cell_ids = getCellsFromTriggerCell(trigger_cell_det_id);
+  // Scintillator
+  if (det == DetId::HGCalHSc) {
+    for (const auto& cell : cell_ids) {
+      triggerCellVector += hscGeometry()->getPosition(cell).basicVector();
+    }
+  }
+  // HFNose
+  else if (det == DetId::HGCalTrigger and
+           HGCalTriggerDetId(trigger_cell_det_id).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
+    for (const auto& cell : cell_ids) {
+      HFNoseDetId cellDetId(cell);
+      triggerCellVector += noseGeometry()->getPosition(cellDetId).basicVector();
+    }
+  }
+  // Silicon
+  else {
+    for (const auto& cell : cell_ids) {
+      HGCSiliconDetId cellDetId(cell);
+      triggerCellVector += (cellDetId.det() == DetId::HGCalEE ? eeGeometry()->getPosition(cellDetId)
+                                                              : hsiGeometry()->getPosition(cellDetId))
+                               .basicVector();
+    }
+  }
+  return GlobalPoint(triggerCellVector / cell_ids.size());
+}
+
+GlobalPoint HGCalTriggerGeometryV9Imp3::getModulePosition(const unsigned module_det_id) const {
+  unsigned det = DetId(module_det_id).det();
+  // Position: barycenter of the module.
+  Basic3DVector<float> moduleVector(0., 0., 0.);
+  const auto cell_ids = getCellsFromModule(module_det_id);
+  // Scintillator
+  if (det == DetId::HGCalHSc) {
+    for (const auto& cell : cell_ids) {
+      moduleVector += hscGeometry()->getPosition(cell).basicVector();
+    }
+  }
+  // HFNose
+  else if (det == DetId::Forward && DetId(module_det_id).subdetId() == ForwardSubdetector::HFNose) {
+    for (const auto& cell : cell_ids) {
+      HFNoseDetId cellDetId(cell);
+      moduleVector += noseGeometry()->getPosition(cellDetId).basicVector();
+    }
+  }  // Silicon
+  else {
+    for (const auto& cell : cell_ids) {
+      HGCSiliconDetId cellDetId(cell);
+      moduleVector += (cellDetId.det() == DetId::HGCalEE ? eeGeometry()->getPosition(cellDetId)
+                                                         : hsiGeometry()->getPosition(cellDetId))
+                          .basicVector();
+    }
+  }
+
+  return GlobalPoint(moduleVector / cell_ids.size());
+}
+
+void HGCalTriggerGeometryV9Imp3::fillMaps() {
+  // read json mapping file
+  json mapping_config_;
+  std::ifstream json_input_file(jsonMappingFile_.fullPath());
+  if (!json_input_file.is_open()) {
+    throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TMapping file\n";
+  }
+  json_input_file >> mapping_config_;
+
+  //Stage 1 to Stage 2 mapping
+  for (unsigned stage1_id = 0; stage1_id < mapping_config_["Stage1"].size(); stage1_id++) {
```

Comment:
Here and below. Should we add some basic checks? For instance that the "Stage1" field is indeed existing in the json.

Reply:
I have now added some `try,catch` statements for each block, which seems to work in testing.

Reply:
This also enabled me to find a couple of bugs in this read-in, which are now fixed.